Fix idempotency for Unix permissions in zip files.

chriskarel opened the pull request:
This fix prevents the unarchive module from reporting 'changed' when a zipfile contains items with Unix permissions that differ from the system default.

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Fixes #22254
Currently, unarchive will always report changed on any items with high non-default Unix permissions.  This is because the module always subtracts the umask from from any permission set it is evaluating, so 'high' permissions sets, or high umask values, will result in the check always failing.  

Example:  A Umask of 022 will result in files with user or group permissions of '6' or '7' (eg: 706, 770) always being marked as changing, because the existing permissions differ when the UMask is applied.  (The examples are checked against 704 and 750)

This PR simply zeros out the UMask value for files with Unix style permissions.  This results in the comparison being done against the intended permissions set.  My testing indicates it preserves idempotence for archives containing files of Unix permissions, Windows permissions, and both.


<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
Unarchive module

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Sep  5 2016, 02:30:38) [GCC 4.8.5 20150623 (Red Hat 4.8.5-9)]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.


  -->
--For Dennis.  I love you little buddy.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

chriskarel committed:
Fix idempotency for Unix permissions in zip files.

This fix prevents the unarchive module from reporting 'changed' when a zipfile contains items with Unix permissions that differ from the system default.

dagwieers commented:
I approve of your change, but we also need integration tests for this eventually.

shipit

chriskarel committed:
Update zip unarchive tests.

Additional tests for the unarchive module with zip files:
- Test file in zip archive with non-default permissions
- Test file added to zip archive with Windows permissions

chriskarel committed:
Typo

chriskarel committed:
Changed names.

Removed dash from new register variable.
-k flag changes file names, converting to match.

chriskarel committed:
Reverting

chriskarel committed:
Additional fix for mixed win/unix archives.

  Turns out my original fix fails under some mixed archives, as setting the umask to zero can be applied to those files.  This creates a per-file umask variable, so a mix of permission types don't cause problems.

chriskarel commented:
OK, I wrote some additional CI tests for this issue, and in doing so:

A) Accidentally committed them into this PR.  I think I've reverted them, but it obviously made things messy.
B) Realized that my initial fix caused regressions under some mixed-content zip files.  I've made additional changes to fix this issue.

Let me know if I need to recreate this PR, or if there are any further steps to get it merged.

dagwieers commented:
The integration tests should be part of this PR if you have them.

chriskarel committed:
CI Checks

CI checks for archives with:
* non default Unix permissions
* Windows permissions

chriskarel committed:
Update main.yml

chriskarel committed:
Debugging

chriskarel committed:
Update unarchive.py

mattclay commented:
CI failure in integration tests on multiple platforms. Here's one example:

https://app.shippable.com/github/ansible/ansible/runs/21886/26/tests

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Shameless bug workaround.

chriskarel committed:
Update main.yml

chriskarel committed:
Workaround for BSD differences.

Using Zipinfo due to lack of support in BSD unzip.
Permissions handling is also different in BSD -- always applies UMASK to file permissions.

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update main.yml

chriskarel committed:
Update unarchive.py

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel commented:
Good grief, what a rabbit hole I've gone down on this one.

The initial issue of #22254 is fixed.

However, BSD's unzip acts completely opposite to all other versions when it comes to permissions and UMask -- it always applies the umask value, overriding what is specified in the archive.  So there's now a test for a BSD system, with a change in behavior.

In writing the Unit tests, I discovered that the versions of Unzip used by Fedora and Ubuntu have a bug that impacts the timestamps on extracted files.  To work around that, I've had to apply specific timestamps to test files.  (That doesn't trigger the bug, no idea why)  The extraneous touch can be removed when the underlying bugs are fixed.  ([Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1451953), [Ubuntu](https://bugs.launchpad.net/ubuntu/+source/unzip/+bug/1691636))

The idempotence test also exposed the fact that BSD's unzip doesn't support listing timestamps in zip archives.  So the unarchive module was configured to use zipinfo instead.  And the CI script was modified to make sure the unzip package is installed (Which wasn't, on BSD).

This PR is now obviously a cavalcade of commit nightmares.  I'm more than happy to recreate this if anyone objects to the mess.  Otherwise, it will just stand as a testament to how much of an iceberg this bug became.  But, if there's anything else I should do, just let me know.

chriskarel committed:
Permission and SELinux changes

chriskarel committed:
Added checks for creating directories and SSH keys for existing users.

chriskarel committed:
Revert "Permission and SELinux changes"

This reverts commit 3c27d95f5d2bf6f8d11003257496aeecde364f7e.

Wrong branch.

chriskarel committed:
Revert "Added checks for creating directories and SSH keys for existing users."

This reverts commit 35833b192a1cf3688821f75b99f432b51207715f.

Wrong branch

chriskarel commented:
Hey, @dagwieers, what more do I need to do to get this approved?  Are we waiting on @pileofrogs as a second `shipit` approver?

dagwieers commented:
Yes, or one of the core members. You may want to bring this up on one of the core meetings...

sivel commented on lib/ansible/modules/files/unarchive.py:
```diff
@@ -391,13 +395,24 @@ def is_unarchived(self):
                 else:
                     permstr = 'rw-rw-rw-'
+                file_umask=umask
+            elif 'bsd' in 'bsd' in systemtype.lower():
```
This looks like a typo.  I think you just want `elif 'bsd' in systemtype.lower()`

chriskarel replied:
Sure is.  Good catch.  Fixed.

chriskarel committed:
Typo fix.

dagwieers commented:
@chriskarel Welcome to my world :-/ I actually wanted to ditch FreeBSD at one point because of how hard it was to keep supporting it. But the main mess comes from the fact we are depending on an external tool and interface via it's stdout and stderr and that's very prone to errors.

In fact I think we should work towards a new pure-python **uncompress** module. It's going to be much easier to support.

PS The fact that FreeBSD would be overwriting files every time is not that bad a behaviour, it used to be the normal behaviour for all platforms a year ago ;-)

abadger committed:
Pep8 fix and formatting from rebase

abadger commented:
I've rebased this PR for you.  Waiting on shippable tests.  If that passes I'll squash and merge.

abadger committed:
One more pep8 fix

abadger commented:
This passed -- One of the centos builds failed due to the yum repository being unavailable and then succeeded on the retry.

abadger commented:
Okay, merged to devel for the 2.4.0 release.   I'm not sure about whether it should be backported for a possible 2.3.3 release or not.  @dagwieers if you think it is safe and should be backported, let me know on IRC and I'll cherry-pick it.
